src/Prometheus/Storage/APCu.php

```diff
@@ -0,0 +1,310 @@
+<?php
+
+
+namespace Prometheus\Storage;
+
+
+use APCUIterator;
+use function class_exists;
+use Prometheus\MetricFamilySamples;
+use RuntimeException;
+
+class APCu implements Adapter
+{
+    const PROMETHEUS_PREFIX = 'prom';
+
+    public function __construct()
+    {
+        if (!extension_loaded('apcu')) {
+            throw new RuntimeException(
+                'The `apcu` extension was not found, please use the APC adapter for earlier PHP versions'
+            );
+        }
+
+        if (!class_exists(APCUIterator::class)) {
+            throw new RuntimeException(
+                'The `apcu` extension version should be 5.0+, please use the APC adapter for earlier versions'
+            );
+        }
+    }
+
+    /**
+     * @return MetricFamilySamples[]
+     */
+    public function collect()
+    {
+        $metrics = $this->collectHistograms();
+        $metrics = array_merge($metrics, $this->collectGauges());
+        $metrics = array_merge($metrics, $this->collectCounters());
+        return $metrics;
+    }
+
+    public function updateHistogram(array $data)
+    {
+        // Initialize the sum
+        $sumKey = $this->histogramBucketValueKey($data, 'sum');
+        $new = apcu_add($sumKey, $this->toInteger(0));
+
+        // If sum does not exist, assume a new histogram and store the metadata
+        if ($new) {
+            apcu_store($this->metaKey($data), json_encode($this->metaData($data)));
+        }
+
+        // Atomically increment the sum
+        // Taken from https://github.com/prometheus/client_golang/blob/66058aac3a83021948e5fb12f1f408ff556b9037/prometheus/value.go#L91
+        $done = false;
+        while (!$done) {
+            $old = apcu_fetch($sumKey);
+            $done = apcu_cas($sumKey, $old, $this->toInteger($this->fromInteger($old) + $data['value']));
+        }
+
+        // Figure out in which bucket the observation belongs
+        $bucketToIncrease = '+Inf';
+        foreach ($data['buckets'] as $bucket) {
+            if ($data['value'] <= $bucket) {
+                $bucketToIncrease = $bucket;
+                break;
+            }
+        }
+
+        // Initialize and increment the bucket
+        apcu_add($this->histogramBucketValueKey($data, $bucketToIncrease), 0);
+        apcu_inc($this->histogramBucketValueKey($data, $bucketToIncrease));
+    }
+
+    public function updateGauge(array $data)
+    {
+        $valueKey = $this->valueKey($data);
+        if ($data['command'] == Adapter::COMMAND_SET) {
+            apcu_store($valueKey, $this->toInteger($data['value']));
+            apcu_store($this->metaKey($data), json_encode($this->metaData($data)));
+        } else {
```

Comment:
Maybe we can simplify this method a bit using early returns?

Reply:
I simply copied over the APC adapter and adjusted it to APCu. This should be enough to satisfy the criteria of the library, I can follow up on a later PR with a full refactoring.